buildSrc/src/main/java/net/neoforged/neodev/GenerateAccessTransformers.java

```diff
@@ -0,0 +1,245 @@
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+import java.util.zip.ZipFile;
+
+public abstract class GenerateAccessTransformers extends DefaultTask {
+    public static final Modifier PUBLIC = mod("public", false, Opcodes.ACC_PUBLIC);
+    public static final Modifier PROTECTED = mod("protected", false, Opcodes.ACC_PUBLIC, Opcodes.ACC_PROTECTED);
+
+    @InputFile
+    public abstract RegularFileProperty getInput();
+
+    @OutputFile
+    public abstract RegularFileProperty getAccessTransformer();
+
+    @Input
+    public abstract ListProperty<AtGroup> getGroups();
+
+    @TaskAction
+    public void exec() throws IOException {
+        Map<String, ClassInfo> targets = new HashMap<>();
+        var visitor = visitor(targets);
+
+        try (var zip = new ZipFile(getInput().getAsFile().get())) {
+            var entries = zip.entries();
+            while (entries.hasMoreElements()) {
+                var next = entries.nextElement();
+                if (next.isDirectory() || !next.getName().endsWith(".class")) continue;
+
+                var lastInner = next.getName().lastIndexOf("$");
+                if (lastInner >= 0) {
+                    // Skip anonymous classes
+                    if (Character.isDigit(next.getName().charAt(lastInner + 1))) continue;
+                }
+
+                try (var in = zip.getInputStream(next)) {
+                    var reader = new ClassReader(in);
+                    reader.accept(visitor, ClassReader.SKIP_CODE | ClassReader.SKIP_DEBUG | ClassReader.SKIP_FRAMES);
+                }
+            }
+        }
+
+        var groupList = getGroups().get();
+        List<String>[] groups = IntStream.range(0, groupList.size())
+                .mapToObj(i -> new ArrayList<String>()).toArray(List[]::new);
+
+        for (ClassInfo value : targets.values()) {
+            for (int i = 0; i < groupList.size(); i++) {
+                var group = groupList.get(i);
+                if (group.classMatch.test(value)) {
+                    if (group.fieldMatch != null) {
+                        for (var field : value.fields) {
+                            if (group.fieldMatch.test(field) && !group.modifier.test(field.access)) {
+                                groups[i].add(group.modifier + " " + value.name.replace('/', '.') + " " + field.name);
```

Comment:
Not a fan of the implicit `toString()` call here for the modifier. Isn't this more readable if it explicitly references a field in the modifier?